src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -0,0 +1,536 @@
+                            "with the variables we will insert for each "
+                            "PSyData region ('{1}*').".
+                            format(routine_name, self.psy_data_var))
+        # Flag that we have now checked for name clashes so that if there's
+        # more than one PSyData node we don't fall over on the symbols
+        # we've previous inserted.
+        self.root.psy_data_name_clashes_checked = True
+
+        # Create a name for this region by finding where this PSyDataNode
+        # is in the list of PSyDataNodes in this Invoke.
+        sched = self.root
+        pnodes = sched.walk(PSyDataNode)
+        region_idx = pnodes.index(self)
+        if self._region_name:
+            region_name = self._region_name
+        else:
+            region_name = "r{0}".format(region_idx)
+        var_name = "psy_data{0}".format(region_idx)
+
+        # Create a variable for this PSyData region
+        reader = FortranStringReader(
+            "type(PSyDataType), save :: {0}".format(var_name))
+        # Tell the reader that the source is free format
+        reader.set_format(FortranFormat(True, False))
+        decln = Fortran2003.Type_Declaration_Stmt(reader)
+        spec_part.content.append(decln)
+
+        # Find the parent in the parse tree - first get a pointer to the
+        # AST for the content of this region.
+        content_ast = self.psy_data_body.children[0].ast
+        # Now store the parent of this region
+        # pylint: disable=protected-access
+        fp_parent = content_ast.parent
+        # pylint: enable=protected-access
+        # Find the location of the AST of our first child node in the
+        # list of child nodes of our parent in the fparser parse tree.
+        ast_start_index = object_index(fp_parent.content,
+                                       content_ast)
+        # Do the same for our last child node
+        if self.psy_data_body[-1].ast_end:
+            ast_end = self.psy_data_body[-1].ast_end
+        else:
+            ast_end = self.psy_data_body[-1].ast
+
+        if ast_end.parent is not fp_parent:
+            raise InternalError(
+                "The beginning ({0}) and end ({1}) nodes of the PSyData "
+                "region in the fparser2 parse tree do not have the same "
+                "parent.".format(content_ast, ast_end))
+        ast_end_index = object_index(fp_parent.content, ast_end)
+
+        # Add the PSyData end call
+        reader = FortranStringReader(
+            "CALL {0}%PostEnd".format(var_name))
+        # Tell the reader that the source is free format
+        reader.set_format(FortranFormat(True, False))
+        pecall = Fortran2003.Call_Stmt(reader)
+        pecall.parent = fp_parent
+        fp_parent.content.insert(ast_end_index+1, pecall)
+
+        # Add the PSyData start call
+        reader = FortranStringReader(
+            "CALL {2}%PreStart('{0}', '{1}', 0, 0)".format(
```

Comment:
Am I right in thinking that this means that the update() method only supports the profiling use case for PSyData? That's fine but can we catch attempts to use it in unsupported ways and raise a suitable NotImplementedError?

Reply:
Indeed. I only modified the update() function enough so that all existing tests would pass, so the PSyData implementation of the profiling interface will work e.g. for Nemo (which I believe needs update atm).